### PR TITLE
Invalid type in where for Merge table (logical error)

### PR DESCRIPTION
**Pull request by KochetovNicolai**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by antonio2368** [comment]
Maybe replace LOGICAL_ERROR with `ILLEGAL_TYPE_OF_COLUMN_FOR_FILTER`?
In the same function (`MergeTreeBaseSelectProcessor::transformHeader`) for invalid types we throw `LOGICAL_ERROR` in 2 places, and `ILLEGAL_TYPE_OF_COLUMN_FOR_FILTER` in one so it's a bit inconsistent.

**Comment by KochetovNicolai** [comment]
I think it is a little bit more correct to check types during query analysis, so that we can throw earlier.

**Comment by KochetovNicolai** [comment]
@antonio2368  you won, I will use your approach :)

**Comment by devcrafter** [comment]
> I think it is a little bit more correct to check types during query analysis, so that we can throw earlier.

Sounds about right to me. We can't do it without new analyzer, or it's just not convenient?

**Review by devcrafter — APPROVED** [review]
LGTM with comment

File: tests/queries/0_stateless/01902_table_function_merge_db_repr.sql [hunk 1]
```diff
@@ -62,6 +62,11 @@ SHOW CREATE TABLE 01902_db.t_merge_1;
 SELECT 'SELECT _database, _table, n FROM merge(currentDatabase(), ^t) ORDER BY _database, _table, n';
 SELECT _database, _table, n FROM merge(currentDatabase(), '^t') ORDER BY _database, _table, n;
 
+--fuzzed LOGICAL_ERROR
```
[the remark this answers is not in the document]
Comment (KochetovNicolai) [reply tests/queries/0_stateless/01902_table_function_merge_db_repr.sql:65]: yes

**Comment by KochetovNicolai** [comment]
> We can't do it without new analyzer, or it's just not convenient?

Yes, not so convenient. It is a check inside StorageMerge, which is already deeper level of abstraction.

**Pull request merged**